otter/log/intent.py

```diff
@@ -0,0 +1,76 @@
+"""
+Logging intents and performing functions
+"""
+
+from functools import partial
+
+import attr
+
+from effect import ComposedDispatcher, Effect, TypeDispatcher, perform
+
+from toolz.dicttoolz import merge
+
+
+@attr.s
+class Log(object):
+    """
+    Intent to log message
+    """
+    msg = attr.ib()
+    fields = attr.ib()
+
+
+@attr.s
+class LogErr(object):
+    """
+    Intent to log error
+    """
+    failure = attr.ib()
+    msg = attr.ib()
+    fields = attr.ib()
+
+
+@attr.s
+class BoundFields(object):
+    """
+    Intent that binds log fields to an effect. Any log or err effect
+    found when performing given effect will be expanded with these fields
+    """
+    effect = attr.ib()
+    fields = attr.ib()
+
+
+def with_log(effect, **fields):
+    return Effect(BoundFields(effect, fields))
+
+
+def msg(msg, **fields):
+    return Effect(Log(msg, fields))
+
+
+def err(failure, msg, **fields):
+    return Effect(LogErr(failure, msg, fields))
+
+
+def perform_logging(log, fields, disp, intent, box):
+    all_fields = merge(fields, intent.fields)
+    if type(intent) is Log:
```

Comment:
The fact that all of the lines in this function (other than the first) involve type-checks on the intent gives a pretty strong indication that this should be three different functions registered as performers for their respective intents. In other words, `TypeDispatcher` already does type dispatching for you.